dkg/sync/server.go

```diff
@@ -63,75 +62,76 @@ func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, defHash
 	}
 
 	server.tcpNode.SetStreamHandler(syncProtoID, func(s network.Stream) {
-		defer s.Close()
-
-		buf := bufio.NewReader(s)
-		b := make([]byte, MsgSize)
-
-		// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
-		n, err := buf.Read(b)
-		if err != nil {
-			log.Error(ctx, "Read client msg from stream", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-
-			return
-		}
-
-		// Number of bytes that are read are the most important
-		b = b[:n]
-
-		msg := new(pb.MsgSync)
-		if err := proto.Unmarshal(b, msg); err != nil {
-			log.Error(ctx, "Unmarshal client msg", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-
-			return
-		}
-
-		pID := s.Conn().RemotePeer()
-		log.Debug(ctx, "Message received from client", z.Any("peer", p2p.PeerName(pID)))
-
-		pubkey, err := pID.ExtractPublicKey()
-		if err != nil {
-			log.Error(ctx, "Get client public key", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-		}
-
-		ok, err := pubkey.Verify(defHash, msg.HashSignature)
-		if err != nil {
-			log.Error(ctx, "Verify defHash signature", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-		}
-
-		resp := &pb.MsgSyncResponse{
-			SyncTimestamp: msg.Timestamp,
-			Error:         "",
-		}
-
-		if !ok {
-			resp.Error = "Invalid Signature"
-		}
-
-		resBytes, err := proto.Marshal(resp)
-		if err != nil {
-			log.Error(ctx, "Marshal server response", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-
-			return
-		}
-
-		_, err = s.Write(resBytes)
-		if err != nil {
-			log.Error(ctx, "Send response to client", err)
-			err = s.Reset()
-			log.Error(ctx, "Stream reset", err)
-
-			return
+		for {
+			buf := bufio.NewReader(s)
+			b := make([]byte, MsgSize)
+			// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
+			n, err := buf.Read(b)
+			if err != nil {
+				log.Error(ctx, "Read client msg from stream", err)
+				err = s.Reset()
+				log.Error(ctx, "Stream reset", err)
+
+				return
+			}
+
+			// Number of bytes that are read are the most important
+			b = b[:n]
+
+			msg := new(pb.MsgSync)
+			if err := proto.Unmarshal(b, msg); err != nil {
+				log.Error(ctx, "Unmarshal client msg", err)
+				err = s.Reset()
+				log.Error(ctx, "Stream reset", err)
+
+				return
+			}
+
+			pID := s.Conn().RemotePeer()
+			log.Debug(ctx, "Message received from client", z.Any("server", p2p.PeerName(pID)))
+
+			pubkey, err := pID.ExtractPublicKey()
+			if err != nil {
+				log.Error(ctx, "Get client public key", err)
+				err = s.Reset()
```

Comment:
discussion: these two lines are repeated in each `if err != nil {}` block
```go
err = s.Reset()
log.Error(ctx, "Stream reset", err)
```
Can we do something to prevent the duplication?

Reply:
yeah, will remove these and add a `defer s.Close()` at the beginning of this function